integration-tests/src/tests/client/resharding_v2.rs

```diff
@@ -0,0 +1,925 @@
+        *block_hash,
+        0,
+    )
+}
+
+// Test cross contract calls
+// This test case tests when there are missing chunks in the produced blocks
+// This is to test that all the chunk management logic is correct, e.g. inclusion of outgoing
+// receipts into next chunks
+fn test_missing_chunks(
+    test_env: &mut TestReshardingEnv,
+    p_missing: f64,
+    target_protocol_version: ProtocolVersion,
+    epoch_length: u64,
+) {
+    // setup
+    let new_accounts = setup_test_env_with_cross_contract_txs(test_env, epoch_length);
+
+    // randomly dropping chunks at the first few epochs when sharding splits happens
+    // make sure initial txs (deploy smart contracts) are processed successfully
+    let drop_chunk_condition = DropChunkCondition::new();
+    for _ in 1..3 {
+        test_env.step(&drop_chunk_condition, target_protocol_version);
+    }
+
+    let drop_chunk_condition = DropChunkCondition::with_probability(p_missing);
+    for _ in 3..3 * epoch_length {
+        test_env.step(&drop_chunk_condition, target_protocol_version);
+        let last_height = test_env.env.clients[0].chain.head().unwrap().height;
+        for height in last_height - 3..=last_height {
+            test_env.check_next_block_with_new_chunk(height);
+        }
+    }
+
+    // make sure all included transactions finished processing
+    let drop_chunk_condition = DropChunkCondition::new();
+    for _ in 3 * epoch_length..5 * epoch_length {
+        test_env.step(&drop_chunk_condition, target_protocol_version);
+        let last_height = test_env.env.clients[0].chain.head().unwrap().height;
+        for height in last_height - 3..=last_height {
+            test_env.check_next_block_with_new_chunk(height);
+        }
+    }
+
+    let successful_txs = test_env.check_tx_outcomes(true);
+    let new_accounts: Vec<_> =
+        successful_txs.iter().flat_map(|(tx_hash, _)| new_accounts.get(tx_hash)).collect();
+    test_env.check_accounts(new_accounts);
+
+    test_env.check_resharding_artifacts(0);
+}
+
+// Use resharding setup to run protocol for couple epochs with given probability
+// of missing chunk. In particular, checks that all txs generated in env have
+// final outcome in the end.
+// TODO: remove logical dependency on resharding.
+fn test_latest_protocol_missing_chunks(p_missing: f64, rng_seed: u64) {
+    init_test_logger();
+    let epoch_length = 10;
+    let mut test_env =
+        create_test_env_for_cross_contract_test(PROTOCOL_VERSION, epoch_length, rng_seed);
+    test_missing_chunks(&mut test_env, p_missing, PROTOCOL_VERSION, epoch_length)
+}
+
+// latest protocol
+
+#[test]
+fn test_latest_protocol_missing_chunks_low_missing_prob() {
+    test_latest_protocol_missing_chunks(0.1, 25);
+}
+
+#[test]
+fn test_latest_protocol_missing_chunks_mid_missing_prob() {
+    test_latest_protocol_missing_chunks(0.5, 26);
+}
+
+#[test]
+fn test_latest_protocol_missing_chunks_high_missing_prob() {
+    test_latest_protocol_missing_chunks(0.9, 27);
+}
```

Comment:
It's funny, those tests don't test resharding, they were just using the resharding testing framework.

Reply:
This file was.. an interesting read 🙃

Reply:
Oh, @Longarithm was using the test framework for some other type of tests